<a href="https://colab.research.google.com/github/michiWS1920/ADS20/blob/master/ADS_Programming_Assignment_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Programming Assignment 1
**Data Wrangling**

In this first problem set you will use your data wrangling skills to answer (hopefully) interessting questions regarding the COVID-19 pandemic.


To perform your analysis you have to load the library ```tidyverse```. If you want to use additional libraries you can add them to the following cell:

In [3]:
library(tidyverse)
library(lubridate)

── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.0     ✔ purrr   0.3.4
✔ tibble  3.0.1     ✔ dplyr   0.8.5
✔ tidyr   1.0.3     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.5.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()


Attaching package: ‘lubridate’


The following objects are masked from ‘package:dplyr’:

    intersect, setdiff, union


The following objects are masked from ‘package:base’:

    date, intersect, setdiff, union




## Task 1
**COVID-19 cases worldwide**

For you first task you will work with data provided by the [EU Open Data Portal](https://data.europa.eu/euodp/en/home). We will use a dataset providing information on the geographic distribution of COVID-19 cases worldwide. You can download the .csv file [here](https://data.europa.eu/euodp/en/data/dataset/covid-19-coronavirus-data/resource/260bbbde-2316-40eb-aec3-7cd7bfc2f590) or load it directly into your notebook by using the download link https://opendata.ecdc.europa.eu/covid19/casedistribution/csv .

Your first task is to load the data into a dataframe:

In [1]:
# Insert your code here
covid <- as.data.frame(read.csv(url("https://opendata.ecdc.europa.eu/covid19/casedistribution/csv")))
covid

dateRep    day month year cases deaths countriesAndTerritories geoId
1     17/05/2020 17  5     2020    0   0     Afghanistan             AF   
2     16/05/2020 16  5     2020 1063  32     Afghanistan             AF   
3     15/05/2020 15  5     2020  113   6     Afghanistan             AF   
4     14/05/2020 14  5     2020  259   3     Afghanistan             AF   
5     13/05/2020 13  5     2020  280   5     Afghanistan             AF   
6     12/05/2020 12  5     2020  285   2     Afghanistan             AF   
7     11/05/2020 11  5     2020  369   5     Afghanistan             AF   
8     10/05/2020 10  5     2020  255   6     Afghanistan             AF   
9     09/05/2020  9  5     2020  215   3     Afghanistan             AF   
10    08/05/2020  8  5     2020  171   2     Afghanistan             AF   
11    07/05/2020  7  5     2020  168   9     Afghanistan             AF   
12    06/05/2020  6  5     2020  330   5     Afghanistan             AF   
13    05/05/2020  5  5     2020  190   5     Afghanistan             AF   
14    04/05/2020  4  5     2020  235  13     Afghanistan             AF   
15    03/05/2020  3  5     2020  134   4     Afghanistan             AF   
16    02/05/2020  2  5     2020  164   4     Afghanistan             AF   
17    01/05/2020  1  5     2020  222   4     Afghanistan             AF   
18    30/04/2020 30  4     2020  122   0     Afghanistan             AF   
19    29/04/2020 29  4     2020  124   3     Afghanistan             AF   
20    28/04/2020 28  4     2020  172   0     Afghanistan             AF   
21    27/04/2020 27  4     2020   68  10     Afghanistan             AF   
22    26/04/2020 26  4     2020  112   4     Afghanistan             AF   
23    25/04/2020 25  4     2020   70   1     Afghanistan             AF   
24    24/04/2020 24  4     2020  105   2     Afghanistan             AF   
25    23/04/2020 23  4     2020   84   4     Afghanistan             AF   
26    22/04/2020 22  4     2020   61   1     Afghanistan             AF   
27    21/04/2020 21  4     2020   35   2     Afghanistan             AF   
28    20/04/2020 20  4     2020   88   3     Afghanistan             AF   
29    19/04/2020 19  4     2020   63   0     Afghanistan             AF   
30    18/04/2020 18  4     2020   51   1     Afghanistan             AF   
⋮     ⋮          ⋮   ⋮     ⋮    ⋮     ⋮      ⋮                       ⋮    
17547 19/04/2020 19  4     2020 1     0      Zimbabwe                ZW   
17548 18/04/2020 18  4     2020 0     0      Zimbabwe                ZW   
17549 17/04/2020 17  4     2020 1     0      Zimbabwe                ZW   
17550 16/04/2020 16  4     2020 6     0      Zimbabwe                ZW   
17551 15/04/2020 15  4     2020 0     0      Zimbabwe                ZW   
17552 14/04/2020 14  4     2020 3     0      Zimbabwe                ZW   
17553 13/04/2020 13  4     2020 0     0      Zimbabwe                ZW   
17554 12/04/2020 12  4     2020 3     0      Zimbabwe                ZW   
17555 11/04/2020 11  4     2020 0     0      Zimbabwe                ZW   
17556 10/04/2020 10  4     2020 0     1      Zimbabwe                ZW   
17557 09/04/2020  9  4     2020 1     1      Zimbabwe                ZW   
17558 08/04/2020  8  4     2020 1     0      Zimbabwe                ZW   
17559 07/04/2020  7  4     2020 0     0      Zimbabwe                ZW   
17560 06/04/2020  6  4     2020 0     0      Zimbabwe                ZW   
17561 05/04/2020  5  4     2020 0     0      Zimbabwe                ZW   
17562 04/04/2020  4  4     2020 1     0      Zimbabwe                ZW   
17563 03/04/2020  3  4     2020 0     0      Zimbabwe                ZW   
17564 02/04/2020  2  4     2020 0     0      Zimbabwe                ZW   
17565 01/04/2020  1  4     2020 1     0      Zimbabwe                ZW   
17566 31/03/2020 31  3     2020 0     0      Zimbabwe                ZW   
17567 30/03/2020 30  3     2020 0     0      Zimbabwe                ZW   
17568 29/03/2020 29  3     2020

You should always familiarize yourself with the dataset at hand befor you do calculations and try to gain insights.

Use functions such as ```head()```, ```sample_n()```, ```summary()```, etc. to deepen your understanding of the available data and briefly describe the dataset (e.g., number of observations, sorting, variables, ...):

In [0]:
# Insert your code here
covid %>%
head()

dateRep    day month year cases deaths countriesAndTerritories geoId
1 15/05/2020 15  5     2020 113   6      Afghanistan             AF   
2 14/05/2020 14  5     2020 259   3      Afghanistan             AF   
3 13/05/2020 13  5     2020 280   5      Afghanistan             AF   
4 12/05/2020 12  5     2020 285   2      Afghanistan             AF   
5 11/05/2020 11  5     2020 369   5      Afghanistan             AF   
6 10/05/2020 10  5     2020 255   6      Afghanistan             AF   
  countryterritoryCode popData2018 continentExp
1 AFG                  37172386    Asia        
2 AFG                  37172386    Asia        
3 AFG                  37172386    Asia        
4 AFG                  37172386    Asia        
5 AFG                  37172386    Asia        
6 AFG                  37172386    Asia

In [0]:
covid %>%
sample_n(10) %>%
summary()

       dateRep       day           month           year          cases       
 25/04/2020:2   Min.   : 2.0   Min.   :1.00   Min.   :2020   Min.   :  0.00  
 02/05/2020:1   1st Qu.: 5.5   1st Qu.:3.25   1st Qu.:2020   1st Qu.:  2.25  
 04/01/2020:1   Median :15.0   Median :4.00   Median :2020   Median : 23.00  
 05/04/2020:1   Mean   :15.2   Mean   :3.50   Mean   :2020   Mean   : 54.10  
 07/04/2020:1   3rd Qu.:25.0   3rd Qu.:4.00   3rd Qu.:2020   3rd Qu.: 37.75  
 09/04/2020:1   Max.   :29.0   Max.   :5.00   Max.   :2020   Max.   :312.00  
 (Other)   :3                                                                
     deaths     countriesAndTerritories     geoId   countryterritoryCode
 Min.   : 0.0   Algeria:1               AT     :1   AUT    :1           
 1st Qu.: 0.0   Austria:1               BT     :1   BTN    :1           
 Median : 2.5   Bhutan :1               DK     :1   DNK    :1           
 Mean   : 3.5   Denmark:1               DZ     :1   DZA    :1           
 3rd Qu.: 5

In [0]:
covid %>%
distinct(countriesAndTerritories)

countriesAndTerritories          
1   Afghanistan                      
2   Albania                          
3   Algeria                          
4   Andorra                          
5   Angola                           
6   Anguilla                         
7   Antigua_and_Barbuda              
8   Argentina                        
9   Armenia                          
10  Aruba                            
11  Australia                        
12  Austria                          
13  Azerbaijan                       
14  Bahamas                          
15  Bahrain                          
16  Bangladesh                       
17  Barbados                         
18  Belarus                          
19  Belgium                          
20  Belize                           
21  Benin                            
22  Bermuda                          
23  Bhutan                           
24  Bolivia                          
25  Bonaire, Saint Eustatius and Saba
26  Bosnia_and_Herzegovina           
27  Botswana                         
28  Brazil                           
29  British_Virgin_Islands           
30  Brunei_Darussalam                
⋮   ⋮                                
181 Sri_Lanka                        
182 Sudan                            
183 Suriname                         
184 Sweden                           
185 Switzerland                      
186 Syria                            
187 Taiwan                           
188 Tajikistan                       
189 Thailand                         
190 Timor_Leste                      
191 Togo                             
192 Trinidad_and_Tobago              
193 Tunisia                          
194 Turkey                           
195 Turks_and_Caicos_islands         
196 Uganda                           
197 Ukraine                          
198 United_Arab_Emirates             
199 United_Kingdom                   
200 United_Republic_of_Tanzania      
201 United_States_of_America         
202 United_States_Virgin_Islands     
203 Uruguay                          
204 Uzbekistan                       
205 Venezuela                        
206 Vietnam                          
207 Western_Sahara                   
208 Yemen                            
209 Zambia                           
210 Zimbabwe

In [0]:
covid %>%
arrange(desc(dateRep)) %>%
head()

dateRep    day month year cases deaths countriesAndTerritories geoId
1 31/12/2019 31  12    2019 0     0      Afghanistan             AF   
2 31/12/2019 31  12    2019 0     0      Algeria                 DZ   
3 31/12/2019 31  12    2019 0     0      Armenia                 AM   
4 31/12/2019 31  12    2019 0     0      Australia               AU   
5 31/12/2019 31  12    2019 0     0      Austria                 AT   
6 31/12/2019 31  12    2019 0     0      Azerbaijan              AZ   
  countryterritoryCode popData2018 continentExp
1 AFG                  37172386    Asia        
2 DZA                  42228429    Africa      
3 ARM                   2951776    Europe      
4 AUS                  24992369    Oceania     
5 AUT                   8847037    Europe      
6 AZE                   9942334    Europe

In [0]:
covid %>%
arrange(desc(-dateRep)) %>%
sample_n(10)

Warning message in Ops.factor(dateRep):
“‘-’ not meaningful for factors”


dateRep    day month year cases deaths countriesAndTerritories geoId
1  28/03/2020 28  3     2020    1    0    Curaçao                 CW   
2  27/03/2020 27  3     2020    0    0    Tunisia                 TN   
3  02/04/2020  2  4     2020    0    0    Cape_Verde              CV   
4  04/03/2020  4  3     2020    2    0    Bahrain                 BH   
5  29/01/2020 29  1     2020    0    0    Lebanon                 LB   
6  04/04/2020  4  4     2020 6663  932    Spain                   ES   
7  19/01/2020 19  1     2020    0    0    Iraq                    IQ   
8  06/04/2020  6  4     2020  387   28    Sweden                  SE   
9  29/04/2020 29  4     2020    0    0    Benin                   BJ   
10 24/03/2020 24  3     2020   35    0    Argentina               AR   
   countryterritoryCode popData2018 continentExp
1  CUW                    159849    America     
2  TUN                  11565204    Africa      
3  CPV                    543767    Africa      
4  BHR                   1569439    Asia        
5  LBN                   6848925    Asia        
6  ESP                  46723749    Europe      
7  IRQ                  38433600    Asia        
8  SWE                  10183175    Europe      
9  BEN                  11485048    Africa      
10 ARG                  44494502    America

*Describe the dataset here*

o deepen your understanding of the available data and briefly describe the dataset (e.g., number of observations, sorting, variables, ...):

Dataset shows data regarding covid cases. 
* Counts new cases/deaths per day per country
* Shows population data for said country/continent
* 17158 rows of data
* sorted by country and date
* Starting with 31/12/2019 until today (15/5/2020)
* includes 210 different countries
* 11 different variables - date in D/M/Y format. 3 columns for day/month/year split up - Cases - Deaths - Country - GeoID - CountryCode - Population data from 2018 and the Continent


Next, you should filter and sort the dataset to answer the following questions:

1.   Which 5 days had the highest number of reported new cases in Germany?
2.   Which 5 days had the highest number of reported deaths in Brazil?

In [0]:
# Insert your code here
covid %>%
filter(countriesAndTerritories == "Germany") %>%
arrange(desc(cases)) %>%
head(5)

dateRep    day month year cases deaths countriesAndTerritories geoId
1 28/03/2020 28  3     2020 6294   72    Germany                 DE   
2 03/04/2020  3  4     2020 6174  145    Germany                 DE   
3 02/04/2020  2  4     2020 6156  140    Germany                 DE   
4 04/04/2020  4  4     2020 6082  141    Germany                 DE   
5 20/03/2020 20  3     2020 5940   30    Germany                 DE   
  countryterritoryCode popData2018 continentExp
1 DEU                  82927922    Europe      
2 DEU                  82927922    Europe      
3 DEU                  82927922    Europe      
4 DEU                  82927922    Europe      
5 DEU                  82927922    Europe

In [0]:
covid %>%
filter(countriesAndTerritories == "Brazil") %>%
arrange(desc(deaths)) %>%
head(5)

dateRep    day month year cases deaths countriesAndTerritories geoId
1 13/05/2020 13  5     2020  9258 881    Brazil                  BR   
2 15/05/2020 15  5     2020 13944 844    Brazil                  BR   
3 09/05/2020  9  5     2020 10222 751    Brazil                  BR   
4 14/05/2020 14  5     2020 11385 749    Brazil                  BR   
5 10/05/2020 10  5     2020 10611 730    Brazil                  BR   
  countryterritoryCode popData2018 continentExp
1 BRA                  209469333   America     
2 BRA                  209469333   America     
3 BRA                  209469333   America     
4 BRA                  209469333   America     
5 BRA                  209469333   America

In [0]:
covid %>%
filter(countriesAndTerritories == "United_States_of_America") %>%
arrange(desc(cases)) %>%
head(5)

dateRep    day month year cases deaths countriesAndTerritories  geoId
1 26/04/2020 26  4     2020 48529 2172   United_States_of_America US   
2 22/04/2020 22  4     2020 37289 2524   United_States_of_America US   
3 11/04/2020 11  4     2020 35527 2087   United_States_of_America US   
4 05/04/2020  5  4     2020 34272 1344   United_States_of_America US   
5 02/05/2020  2  5     2020 33955 2062   United_States_of_America US   
  countryterritoryCode popData2018 continentExp
1 USA                  327167434   America     
2 USA                  327167434   America     
3 USA                  327167434   America     
4 USA                  327167434   America     
5 USA                  327167434   America

Having established first insights, you now have to aggregate and transform the dataset for further analysis. 

First, calculate the number of total confirmed cases and the number of total deaths for each country. Which countries are hit hardest in terms of total cases and total deaths? 

In [0]:
covid %>%
group_by(countriesAndTerritories) %>%
summarise(totalcases = sum(cases), totaldeaths = sum(deaths)) %>%
arrange(-totalcases) -> covid_sum
covid_sum %>%
head()

countriesAndTerritories  totalcases totaldeaths
1 United_States_of_America 1417889    85906      
2 Russia                    252245     2305      
3 United_Kingdom            233151    33614      
4 Spain                     229540    27321      
5 Italy                     223096    31368      
6 Brazil                    202918    13993

In [0]:
covid_sum %>%
arrange(-totaldeaths) %>%
head()

countriesAndTerritories  totalcases totaldeaths
1 United_States_of_America 1417889    85906      
2 United_Kingdom            233151    33614      
3 Italy                     223096    31368      
4 France                    141356    27425      
5 Spain                     229540    27321      
6 Brazil                    202918    13993

Next, we are interested in the monthly [incidence rate](https://en.wikipedia.org/wiki/Incidence_(epidemiology)) on a continent level. To calculate this rate, you first have to summarize the dataset on a year, month and continent level. Subsequently, you can calculate the incidence rate by using the ```mutate()``` function.

Hint 1: Some countries do not report population data. To ignore NA values you can use the parameter ```na.rm=TRUE``` in most dplyr functions.

Hint 2: You can remove the continet ```other``` prior to your analysis.

![alt text](https://i.imgur.com/BdeWYT8.png)

In [4]:
#filter out Other
covid %>%
filter(continentExp != "Other") -> covid_continent
covid_continent %>%
distinct(continentExp)

continentExp
1 Asia        
2 Europe      
3 Africa      
4 America     
5 Oceania

In [69]:
#replace NA with 0
covid_continent[is.na(covid_continent)] <- 0
covid_continent

Warning message in `[<-.factor`(`*tmp*`, thisvar, value = 0):
“invalid factor level, NA generated”


dateRep    day month year cases deaths countriesAndTerritories geoId
1     17/05/2020 17  5     2020    0   0     Afghanistan             AF   
2     16/05/2020 16  5     2020 1063  32     Afghanistan             AF   
3     15/05/2020 15  5     2020  113   6     Afghanistan             AF   
4     14/05/2020 14  5     2020  259   3     Afghanistan             AF   
5     13/05/2020 13  5     2020  280   5     Afghanistan             AF   
6     12/05/2020 12  5     2020  285   2     Afghanistan             AF   
7     11/05/2020 11  5     2020  369   5     Afghanistan             AF   
8     10/05/2020 10  5     2020  255   6     Afghanistan             AF   
9     09/05/2020  9  5     2020  215   3     Afghanistan             AF   
10    08/05/2020  8  5     2020  171   2     Afghanistan             AF   
11    07/05/2020  7  5     2020  168   9     Afghanistan             AF   
12    06/05/2020  6  5     2020  330   5     Afghanistan             AF   
13    05/05/2020  5  5     2020  190   5     Afghanistan             AF   
14    04/05/2020  4  5     2020  235  13     Afghanistan             AF   
15    03/05/2020  3  5     2020  134   4     Afghanistan             AF   
16    02/05/2020  2  5     2020  164   4     Afghanistan             AF   
17    01/05/2020  1  5     2020  222   4     Afghanistan             AF   
18    30/04/2020 30  4     2020  122   0     Afghanistan             AF   
19    29/04/2020 29  4     2020  124   3     Afghanistan             AF   
20    28/04/2020 28  4     2020  172   0     Afghanistan             AF   
21    27/04/2020 27  4     2020   68  10     Afghanistan             AF   
22    26/04/2020 26  4     2020  112   4     Afghanistan             AF   
23    25/04/2020 25  4     2020   70   1     Afghanistan             AF   
24    24/04/2020 24  4     2020  105   2     Afghanistan             AF   
25    23/04/2020 23  4     2020   84   4     Afghanistan             AF   
26    22/04/2020 22  4     2020   61   1     Afghanistan             AF   
27    21/04/2020 21  4     2020   35   2     Afghanistan             AF   
28    20/04/2020 20  4     2020   88   3     Afghanistan             AF   
29    19/04/2020 19  4     2020   63   0     Afghanistan             AF   
30    18/04/2020 18  4     2020   51   1     Afghanistan             AF   
⋮     ⋮          ⋮   ⋮     ⋮    ⋮     ⋮      ⋮                       ⋮    
17483 19/04/2020 19  4     2020 1     0      Zimbabwe                ZW   
17484 18/04/2020 18  4     2020 0     0      Zimbabwe                ZW   
17485 17/04/2020 17  4     2020 1     0      Zimbabwe                ZW   
17486 16/04/2020 16  4     2020 6     0      Zimbabwe                ZW   
17487 15/04/2020 15  4     2020 0     0      Zimbabwe                ZW   
17488 14/04/2020 14  4     2020 3     0      Zimbabwe                ZW   
17489 13/04/2020 13  4     2020 0     0      Zimbabwe                ZW   
17490 12/04/2020 12  4     2020 3     0      Zimbabwe                ZW   
17491 11/04/2020 11  4     2020 0     0      Zimbabwe                ZW   
17492 10/04/2020 10  4     2020 0     1      Zimbabwe                ZW   
17493 09/04/2020  9  4     2020 1     1      Zimbabwe                ZW   
17494 08/04/2020  8  4     2020 1     0      Zimbabwe                ZW   
17495 07/04/2020  7  4     2020 0     0      Zimbabwe                ZW   
17496 06/04/2020  6  4     2020 0     0      Zimbabwe                ZW   
17497 05/04/2020  5  4     2020 0     0      Zimbabwe                ZW   
17498 04/04/2020  4  4     2020 1     0      Zimbabwe                ZW   
17499 03/04/2020  3  4     2020 0     0      Zimbabwe                ZW   
17500 02/04/2020  2  4     2020 0     0      Zimbabwe                ZW   
17501 01/04/2020  1  4     2020 1     0      Zimbabwe                ZW   
17502 31/03/2020 31  3     2020 0     0      Zimbabwe                ZW   
17503 30/03/2020 30  3     2020 0     0      Zimbabwe                ZW   
17504 29/03/2020 29  3     2020

In [0]:
#new column for continentPop
covid_continent %>%
group_by(continentExp, popData2018) %>%
mutate(ContinentPop = sum(popData2018)) -> covid_prep

In [73]:
# first aggregation to monthly _cases
# forgot to get popData on a continent level
covid_prep %>%
group_by(year, month, continentExp) %>%
mutate(monthly_cases = sum(cases)) -> covid_inc
covid_inc

dateRep    day month year cases deaths countriesAndTerritories geoId
1     17/05/2020 17  5     2020    0   0     Afghanistan             AF   
2     16/05/2020 16  5     2020 1063  32     Afghanistan             AF   
3     15/05/2020 15  5     2020  113   6     Afghanistan             AF   
4     14/05/2020 14  5     2020  259   3     Afghanistan             AF   
5     13/05/2020 13  5     2020  280   5     Afghanistan             AF   
6     12/05/2020 12  5     2020  285   2     Afghanistan             AF   
7     11/05/2020 11  5     2020  369   5     Afghanistan             AF   
8     10/05/2020 10  5     2020  255   6     Afghanistan             AF   
9     09/05/2020  9  5     2020  215   3     Afghanistan             AF   
10    08/05/2020  8  5     2020  171   2     Afghanistan             AF   
11    07/05/2020  7  5     2020  168   9     Afghanistan             AF   
12    06/05/2020  6  5     2020  330   5     Afghanistan             AF   
13    05/05/2020  5  5     2020  190   5     Afghanistan             AF   
14    04/05/2020  4  5     2020  235  13     Afghanistan             AF   
15    03/05/2020  3  5     2020  134   4     Afghanistan             AF   
16    02/05/2020  2  5     2020  164   4     Afghanistan             AF   
17    01/05/2020  1  5     2020  222   4     Afghanistan             AF   
18    30/04/2020 30  4     2020  122   0     Afghanistan             AF   
19    29/04/2020 29  4     2020  124   3     Afghanistan             AF   
20    28/04/2020 28  4     2020  172   0     Afghanistan             AF   
21    27/04/2020 27  4     2020   68  10     Afghanistan             AF   
22    26/04/2020 26  4     2020  112   4     Afghanistan             AF   
23    25/04/2020 25  4     2020   70   1     Afghanistan             AF   
24    24/04/2020 24  4     2020  105   2     Afghanistan             AF   
25    23/04/2020 23  4     2020   84   4     Afghanistan             AF   
26    22/04/2020 22  4     2020   61   1     Afghanistan             AF   
27    21/04/2020 21  4     2020   35   2     Afghanistan             AF   
28    20/04/2020 20  4     2020   88   3     Afghanistan             AF   
29    19/04/2020 19  4     2020   63   0     Afghanistan             AF   
30    18/04/2020 18  4     2020   51   1     Afghanistan             AF   
⋮     ⋮          ⋮   ⋮     ⋮    ⋮     ⋮      ⋮                       ⋮    
17483 19/04/2020 19  4     2020 1     0      Zimbabwe                ZW   
17484 18/04/2020 18  4     2020 0     0      Zimbabwe                ZW   
17485 17/04/2020 17  4     2020 1     0      Zimbabwe                ZW   
17486 16/04/2020 16  4     2020 6     0      Zimbabwe                ZW   
17487 15/04/2020 15  4     2020 0     0      Zimbabwe                ZW   
17488 14/04/2020 14  4     2020 3     0      Zimbabwe                ZW   
17489 13/04/2020 13  4     2020 0     0      Zimbabwe                ZW   
17490 12/04/2020 12  4     2020 3     0      Zimbabwe                ZW   
17491 11/04/2020 11  4     2020 0     0      Zimbabwe                ZW   
17492 10/04/2020 10  4     2020 0     1      Zimbabwe                ZW   
17493 09/04/2020  9  4     2020 1     1      Zimbabwe                ZW   
17494 08/04/2020  8  4     2020 1     0      Zimbabwe                ZW   
17495 07/04/2020  7  4     2020 0     0      Zimbabwe                ZW   
17496 06/04/2020  6  4     2020 0     0      Zimbabwe                ZW   
17497 05/04/2020  5  4     2020 0     0      Zimbabwe                ZW   
17498 04/04/2020  4  4     2020 1     0      Zimbabwe                ZW   
17499 03/04/2020  3  4     2020 0     0      Zimbabwe                ZW   
17500 02/04/2020  2  4     2020 0     0      Zimbabwe                ZW   
17501 01/04/2020  1  4     2020 1     0      Zimbabwe                ZW   
17502 31/03/2020 31  3     2020 0     0      Zimbabwe                ZW   
17503 30/03/2020 30  3     2020 0     0      Zimbabwe                ZW   
17504 29/03/2020 29  3     2020

In [0]:
# nice format, but missing the pop data. If used in group_by the output has duplicates. Need to update popdata for continent level
covid_prep %>%
group_by(year, month, continentExp) %>%
summarise_each(funs(sum), cases) -> covid_grp
#covid_grp


In [81]:
covid_inc %>%
group_by(year, month, continentExp) %>%
mutate(inc_rate = monthly_cases / ContinentPop, inc_rate=scales::percent(inc_rate))

dateRep    day month year cases deaths countriesAndTerritories geoId
1     17/05/2020 17  5     2020    0   0     Afghanistan             AF   
2     16/05/2020 16  5     2020 1063  32     Afghanistan             AF   
3     15/05/2020 15  5     2020  113   6     Afghanistan             AF   
4     14/05/2020 14  5     2020  259   3     Afghanistan             AF   
5     13/05/2020 13  5     2020  280   5     Afghanistan             AF   
6     12/05/2020 12  5     2020  285   2     Afghanistan             AF   
7     11/05/2020 11  5     2020  369   5     Afghanistan             AF   
8     10/05/2020 10  5     2020  255   6     Afghanistan             AF   
9     09/05/2020  9  5     2020  215   3     Afghanistan             AF   
10    08/05/2020  8  5     2020  171   2     Afghanistan             AF   
11    07/05/2020  7  5     2020  168   9     Afghanistan             AF   
12    06/05/2020  6  5     2020  330   5     Afghanistan             AF   
13    05/05/2020  5  5     2020  190   5     Afghanistan             AF   
14    04/05/2020  4  5     2020  235  13     Afghanistan             AF   
15    03/05/2020  3  5     2020  134   4     Afghanistan             AF   
16    02/05/2020  2  5     2020  164   4     Afghanistan             AF   
17    01/05/2020  1  5     2020  222   4     Afghanistan             AF   
18    30/04/2020 30  4     2020  122   0     Afghanistan             AF   
19    29/04/2020 29  4     2020  124   3     Afghanistan             AF   
20    28/04/2020 28  4     2020  172   0     Afghanistan             AF   
21    27/04/2020 27  4     2020   68  10     Afghanistan             AF   
22    26/04/2020 26  4     2020  112   4     Afghanistan             AF   
23    25/04/2020 25  4     2020   70   1     Afghanistan             AF   
24    24/04/2020 24  4     2020  105   2     Afghanistan             AF   
25    23/04/2020 23  4     2020   84   4     Afghanistan             AF   
26    22/04/2020 22  4     2020   61   1     Afghanistan             AF   
27    21/04/2020 21  4     2020   35   2     Afghanistan             AF   
28    20/04/2020 20  4     2020   88   3     Afghanistan             AF   
29    19/04/2020 19  4     2020   63   0     Afghanistan             AF   
30    18/04/2020 18  4     2020   51   1     Afghanistan             AF   
⋮     ⋮          ⋮   ⋮     ⋮    ⋮     ⋮      ⋮                       ⋮    
17483 19/04/2020 19  4     2020 1     0      Zimbabwe                ZW   
17484 18/04/2020 18  4     2020 0     0      Zimbabwe                ZW   
17485 17/04/2020 17  4     2020 1     0      Zimbabwe                ZW   
17486 16/04/2020 16  4     2020 6     0      Zimbabwe                ZW   
17487 15/04/2020 15  4     2020 0     0      Zimbabwe                ZW   
17488 14/04/2020 14  4     2020 3     0      Zimbabwe                ZW   
17489 13/04/2020 13  4     2020 0     0      Zimbabwe                ZW   
17490 12/04/2020 12  4     2020 3     0      Zimbabwe                ZW   
17491 11/04/2020 11  4     2020 0     0      Zimbabwe                ZW   
17492 10/04/2020 10  4     2020 0     1      Zimbabwe                ZW   
17493 09/04/2020  9  4     2020 1     1      Zimbabwe                ZW   
17494 08/04/2020  8  4     2020 1     0      Zimbabwe                ZW   
17495 07/04/2020  7  4     2020 0     0      Zimbabwe                ZW   
17496 06/04/2020  6  4     2020 0     0      Zimbabwe                ZW   
17497 05/04/2020  5  4     2020 0     0      Zimbabwe                ZW   
17498 04/04/2020  4  4     2020 1     0      Zimbabwe                ZW   
17499 03/04/2020  3  4     2020 0     0      Zimbabwe                ZW   
17500 02/04/2020  2  4     2020 0     0      Zimbabwe                ZW   
17501 01/04/2020  1  4     2020 1     0      Zimbabwe                ZW   
17502 31/03/2020 31  3     2020 0     0      Zimbabwe                ZW   
17503 30/03/2020 30  3     2020 0     0      Zimbabwe                ZW   
17504 29/03/2020 29  3     2020

Modify your code to also calculate the death rate on the same aggregation level.

In [0]:
# Insert your code here

## Task 2

**Contact tracing**

Governments rely on contact tracing to avoid the spread of the COVID-19 disease. Once a new case has been identified, health authorities track down people who may have been in contact with the sick person. 

Recently, there has been a new case and you are tasked to help local authorities to identify all contact persons. Initially, your only infomation is that the infected person (Mike) started to show symptoms on May 1, 2020 and was tested positiv and quarantined on the same day. However, infected persons can be contagious for up to 2 weeks without showing any symptoms. Hence, you have to reconstruct all contacts in that time window. 

The health authorities provide you with a dataframe ```df``` containing the personal information on Mike.

In [0]:
df = data.frame(first_name = 'Mike', last_name = 'Williams', job = 'Student', passport_number = 204934343, phone_number = '+1-202-555-0121', quarantine_date = '2020-05-1', stringsAsFactors = FALSE)
df

In order to identify all contacts you first have to find the date Mike stated to be contagious.

Hint: You should use the ```lubridate``` package.

In [0]:
# Insert your code here

During your investigations you find out that Mike visited the local gym once in the last month. Luckily, the gym uses an electronic gate and logs all visits with time stamps and member id. To help you reconstruct the data, the gym owner provides you with an additional file containing information on the members.  

You can downlaod the member data from https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/gym_data.csv
and the visitor logs from https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/gym_log.csv

Load both dataframes and identify all possible contact persons (by name and member number) that could have met Mike in the gym for at least 15 minutes while he was contagious.



Hint: You might have to use ```lubridate``` to transform the checkin and checkout columns


In [0]:
# Insert your code here

Additionally, you learn that Mike took several exams in the university. While the university has social distancing measures in place it is still important to identify all students sitting close to Mike during one of the exams. Students with seat numbers in the range of -5 to +5 to Mike have to be identified. 

You can download a list of all exams Mike was registered for from https://raw.githubusercontent.com/NikoStein/ADS20_Data/master/Assignment%201/exams_data.csv





In [0]:
# Insert your code here

After downloading and inspecting the files you have to find all exams taken by Mike while he was contagious.

In [0]:
# Insert your code here

Next, identify all students that came into contact with Mike during one of the relevant exams. Store the first and last names of the students in a new data frame.

In [0]:
# Insert your code here